datafusion/functions/src/datetime/to_time.rs

```diff
@@ -0,0 +1,403 @@
+                .map(|s| parse_time_with_formats(s, &formats))
+                .transpose()?;
+            Ok(ColumnarValue::Scalar(ScalarValue::Time64Nanosecond(result)))
+        }
+        ColumnarValue::Array(array) => {
+            let result = match array.data_type() {
+                Utf8 => parse_time_array(&array.as_string::<i32>(), &formats)?,
+                LargeUtf8 => parse_time_array(&array.as_string::<i64>(), &formats)?,
+                Utf8View => parse_time_array(&array.as_string_view(), &formats)?,
+                other => return exec_err!("Unsupported type for to_time: {other}"),
+            };
+            Ok(ColumnarValue::Array(Arc::new(result)))
+        }
+        other => exec_err!("Unsupported argument for to_time: {other:?}"),
+    }
+}
+
+/// Collect format strings from arguments, erroring on non-scalar inputs
+fn collect_formats(args: &[ColumnarValue]) -> Result<Vec<&str>> {
+    if args.len() <= 1 {
+        return Ok(DEFAULT_TIME_FORMATS.to_vec());
+    }
+
+    let mut formats = Vec::with_capacity(args.len() - 1);
+    for (i, arg) in args[1..].iter().enumerate() {
+        match arg {
+            ColumnarValue::Scalar(ScalarValue::Utf8(Some(s)))
+            | ColumnarValue::Scalar(ScalarValue::LargeUtf8(Some(s)))
+            | ColumnarValue::Scalar(ScalarValue::Utf8View(Some(s))) => {
+                formats.push(s.as_str());
+            }
+            ColumnarValue::Scalar(ScalarValue::Utf8(None))
+            | ColumnarValue::Scalar(ScalarValue::LargeUtf8(None))
+            | ColumnarValue::Scalar(ScalarValue::Utf8View(None)) => {
+                // Skip null format strings
+            }
+            ColumnarValue::Array(_) => {
+                return exec_err!(
+                    "to_time format argument {} must be a scalar, not an array",
+                    i + 2 // argument position (1-indexed, +1 for the first arg)
+                );
+            }
+            other => {
+                return exec_err!(
+                    "to_time format argument {} has unsupported type: {:?}",
+                    i + 2,
+                    other.data_type()
+                );
+            }
+        }
+    }
+    Ok(formats)
+}
+
+/// Extract time portion from timestamp using Arrow cast kernel
+fn timestamp_to_time(arg: &ColumnarValue) -> Result<ColumnarValue> {
+    let target_type = Time64(arrow::datatypes::TimeUnit::Nanosecond);
+
+    match arg {
+        ColumnarValue::Scalar(scalar) => {
+            // Convert scalar to array of size 1, cast, then extract result
+            let array = scalar.to_array()?;
+            let time_array = cast(&array, &target_type)?;
+            let time_value = ScalarValue::try_from_array(&time_array, 0)?;
+            Ok(ColumnarValue::Scalar(time_value))
+        }
+        ColumnarValue::Array(array) => {
+            let time_array = cast(array, &target_type)?;
+            Ok(ColumnarValue::Array(time_array))
+        }
+    }
```

Comment:
```suggestion
    Ok(arg.cast_to(&Time64(arrow::datatypes::TimeUnit::Nanosecond), None)?)
```

https://docs.rs/datafusion/latest/datafusion/logical_expr/enum.ColumnarValue.html#method.cast_to